src/pudl/transform/eia861.py

```diff
@@ -723,13 +717,39 @@ def _tidy_class_dfs(
     data_cols = data_cols.stack(level=0, dropna=False).reset_index()
     denorm_cols = _filter_non_class_cols(raw_df, class_list).reset_index()
 
+    # Check to make sure that the idx_cols are actually valid primary key cols:
+    # This is tricky, because NA values in the BA Code column creates actual duplicate
+    # values. These NA values are filled with UNK and the duplicates are later dropped,
+    # which means we are losing data. But it's not obvious how we could actually fill
+    # in real BA codes, so we consolidate the indistinguisbale records.
+    # These are only a fraction of a percent of all records, and only affect a couple
+    # of tables. In sales_eia861 there are 67 duplicate records.
+    # See: https://github.com/catalyst-cooperative/pudl/issues/2638
+    dupe_mask = data_cols.duplicated(subset=idx_cols + [class_type], keep=False)
+    dupes = data_cols[dupe_mask]
+    fraction_dupes = len(dupes) / len(data_cols)
+    err_msg = (
+        f"{df_name} table: Found {len(dupes)}/{len(data_cols)} "
+        f"({fraction_dupes:0.2%}) records with duplicated PKs. "
+    )
+    if fraction_dupes <= 0.005:
+        err_msg += "Consolidating duplicated records."
+        logger.info(err_msg)
+        deduped = dupes.groupby(idx_cols + [class_type], as_index=False).sum()
+        data_cols = pd.concat([data_cols[~dupe_mask], deduped], axis="index")
+        # PK dupes also affect denorm_cols. Drop dupes to ensure clean 1 to many merge:
+        denorm_cols = denorm_cols.drop_duplicates(subset=idx_cols)
```

Comment:
I would similarly want to have some kind of check to make sure we aren't dropping huge amounts of data here.

Reply:
Okay, also checking the non-data columns now...